src/io/reader.rs

```diff
@@ -0,0 +1,108 @@
+use std::fs::File;
+use std::io::{self, BufRead, BufReader, Cursor, Read, Seek, SeekFrom};
+use std::path::Path;
+
+use dynimage::DynamicImage;
+use image::ImageFormat;
+use {ImageError, ImageResult};
+
+use super::free_functions;
+
+/// A multi-format image reader.
+///
+/// Wraps an input reader to facilitate automatic detection of an image's format, appropriate
+/// decoding method, and dispatches into the set of supported `ImageDecoder` implementations.
+pub struct Reader<R: Read> {
+    /// The reader.
+    inner: R,
+    /// The format, if one has been set or deduced.
+    format: Option<ImageFormat>,
+}
+
+impl<R: Read> Reader<R> {
+    pub fn new(reader: R) -> Self {
+        Reader {
+            inner: reader,
+            format: None,
+        }
+    }
+
+    /// Construct a reader with specified format.
+    pub fn with_format(reader: R, format: ImageFormat) -> Self {
+        Reader {
+            inner: reader,
+            format: Some(format),
+        }
+    }
+
+    /// Get the currently determined format.
+    pub fn format(&self) -> Option<ImageFormat> {
+        self.format
+    }
+
+    /// Unwrap the reader.
+    pub fn into_inner(self) -> R {
+        self.inner
+    }
+}
+
+impl Reader<BufReader<File>> {
+    /// Open a file to read. Format will be guessed.
+    /// Note: Failure to guess is not an error. The `Err` is only for `File::open`.
+    pub fn open<P>(path: P) -> io::Result<Self> where P: AsRef<Path> {
+        let path = path.as_ref();
+
+        let file = File::open(path)?;
+        let format = ImageFormat::from_path(path);
+
+        Ok(Reader {
+            inner: BufReader::new(file),
+            format: format.ok(),
+        })
+    }
+}
+
+impl<R: BufRead + Seek> Reader<R> {
+    /// Replace the format with a guess based on the content.
+    ///
+    /// If the guess fails, the format is unchanged. Error is for seek, hence io::Error 
+    /// and not ImageError.
+    pub fn guess_format(&mut self) -> io::Result<()> {
+        let mut start = [0; 16];
+
+        // Save current offset, read start, restore offset.
+        let cur = self.inner.seek(SeekFrom::Current(0))?;
+        let len = io::copy(
+            // Accept shorter files but read at most 16 bytes.
+            &mut self.inner.by_ref().take(16),
+            &mut Cursor::new(&mut start[..]))?;
+        self.inner.seek(SeekFrom::Start(cur))?;
+
+        self.format = free_functions::guess_format(&start[..len as usize]).ok();
+        Ok(())
+    }
+
+    /// Read the image dimensions.
+    /// Uses the current format to construct the correct reader.
+    pub fn image_dimensions(self) -> ImageResult<(u32, u32)> {
```

Comment:
I'd rather there be a `into_decoder(self) -> ImageResult<impl ImageDecoder>` function instead. That way after getting the dimensions you don't have to start over to get other pieces of metadata or to decode the image.

Reply:
`into_decoder` needs to dynamically allocate, this one doesn't. That was the rationale at least. 

But yeah, there should be an `into_decoder` in addition.

Reply:
Why is dynamic allocation required? I'm just imagining that function would return an enum over the different decoder types (a Box<dyn ImageDecoder> wouldn't work because the trait isn't object safe).

Reply:
You're right, an enum may be possible.

But even if it is private it will be hard to maintain, if it's a public enum it will be even harder. Each feature flags would add or remove a variant which would mess with a lot of `match` statements. In the end, I fear that it would turn into a similar state as `DynamicImage` currently. Still might be one way forward but I don't see myself adding it to this PR right now, that's something I would like to consider and sketch separately.

Reply:
I actually wrote out the code for a DynamicDecoder enum in #904. It worked out to be about 40 lines total. With some refactoring, that type plus the one case in `into_decoder` could replace nearly all of the other instances where we currently mach over image types.

Reply:
With `type Reader = Box<dyn Read + 'a>` which is not optimal since it could have a similar `enum`. All together, the interface for `DynamicDecoder` is somewhat removed from the interface for `Reader` and we can always replace the internals when it is ready.